pyboolnet/control_strategies.py

```diff
@@ -0,0 +1,486 @@
+        * *max_output*: maximum number of trap spaces computed. Default value: 1000000.
+
+    **returns**:
+        * *answer*: True if the *candidate* is a control strategy for the *target*, False otherwise. When False, a counterexample (minimal trap space disjoint with target or state not satisfying the CTL query) is also returned.
+
+    **example**::
+        >>> is_control_strategy(primes, {'v1': 1}, [{'v2': 0}, {'v3':1}], "asynchronous")
+    """
+
+    # Preliminary setting
+
+    if type(target) == dict:
+        print("The target must be a list of subspaces.")
+        return
+
+    if control_direct_percolation(primes, candidate, target, silent=True):
+        return True
+
+    new_primes = fix_components_and_reduce(primes, candidate, list(set([x for subs in subspaces for x in subs])))
+    tsmin = compute_trap_spaces(new_primes, "min", max_output=max_output)
+
+    # Checking control is valid in the trap spaces
+    if not control_is_valid_in_trap_spaces(new_primes, tsmin, target, update):
+        return False
+
+    # Checking general CTL query
+    return run_control_query(new_primes, target, update)
+
+
+def compute_control_strategies_with_completeness(primes: dict, target: dict, update: str = "asynchronous", limit: int = 3, avoid_nodes: List[str] = None, silent: bool = False, starting_length: int = 0, previous_cs: List[dict] = None, known_cs: List[dict] = None) -> List[dict]:
+    """
+    Identify control strategies for the *target* subspace using completeness.
+
+    **arguments**:
+        *primes*: prime implicants.
+        *target*: target subspace.
+        *update*: the type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+        *limit*: maximal size of the control strategies. Default value: 3.
+        *silent*: if True, it does not print infos to screen. Default value: False.
+        *starting_length*: minimum possible size of the control strategies. Default value: 0.
+        *previous_cs*: list of already identified control strategies. Default value: empty list.
+        *avoid_nodes*: list of nodes that cannot be part of any control strategy. Default value: empty list.
+
+    **returns**:
+        * *cs_total*: list of control strategies for the *target* subspace obtained using completeness.
+
+    **example**::
+        >>> control_strategies_completeness(primes, {'v1': 1}, "asynchronous")
+
+    """
+
+    # Initializing
+
+    if previous_cs is None:
+        previous_cs = []
+    if avoid_nodes is None:
+        avoid_nodes = []
+    if known_cs is None:
+        known_cs = []
+
+    # Preliminary setting
+
+    candidate_variables = [x for x in primes.keys() if x not in avoid_nodes]
+    cs_total = previous_cs
+    perc_true = known_cs
+    perc_false = []
+
+    common_vars_in_cs = find_common_variables_in_cs(primes, [target])
+    candidate_variables = [x for x in primes.keys() if x not in common_vars_in_cs.keys() and x not in avoid_nodes]
+    if not silent:
```

Comment:
use logging

Reply:
Do you mean that instead of:
```
if not silent:
    print("Intervention:", candidate)
```
I should use:
```
log.info("Intervention:", candidate)
```

Reply:
Yes.